libs/core/include/core/bloom_filter.hpp

```diff
@@ -0,0 +1,224 @@
+    HashSourceIterator &operator=(HashSourceIterator &&) noexcept;
+
+    HashSourceIterator() = delete;
+
+    bool operator!=(HashSourceIterator const &other) const;
+    bool operator==(HashSourceIterator const &other) const;
+
+    HashSourceIterator const operator++(int);
+    HashSourceIterator &     operator++();
+
+    std::size_t operator*() const;
+
+  private:
+    explicit HashSourceIterator(HashSource const *source, std::size_t index);
+
+    std::size_t       hash_index_{};
+    HashSource const *source_{};
+
+    friend class HashSource;
+  };
+
+  HashSourceIterator begin() const;
+  HashSourceIterator end() const;
+  HashSourceIterator cbegin() const;
+  HashSourceIterator cend() const;
+
+private:
+  HashSource(HashSourceFactory::Functions const &, HashSourceFactory::Bytes const &);
+
+  std::size_t getHash(std::size_t index) const;
+
+  std::vector<std::size_t> data_;
+
+  friend class HashSourceFactory;
+};
+
+}  // namespace internal
+
+class BloomFilter
+{
+public:
+  using Bytes     = internal::HashSourceFactory::Bytes;
+  using Function  = internal::HashSourceFactory::Function;
+  using Functions = internal::HashSourceFactory::Functions;
+
+  enum class Type
+  {
+    DUMMY,
+    BASIC
+  };
+
+  static std::unique_ptr<BloomFilter> create(Type);
+
+  virtual ~BloomFilter() = default;
+
+  /*
+   * Query the Bloom filter for a given entry.
+   *
+   * Returns false if the entry is definitely absent; true otherwise.
+   */
+  virtual bool Match(Bytes const &) = 0;
+
+  /*
+   * Add a new entry to the filter.
+   */
+  virtual void Add(Bytes const &) = 0;
+
+  /*
+   * Clients may use this to report how many false positives they identified.
+   * This information is used internally by the filter to keep track of the
+   * false positive rate.
+   *
+   * Return false if the filter's measured false positive rate exceeds its
+   * target value and rebuilding the filter may be advisable; true otherwise.
+   */
+  virtual bool ReportFalsePositives(std::size_t) = 0;
+};
+
+class BasicBloomFilter : public BloomFilter
+{
+public:
+  BasicBloomFilter();
+  explicit BasicBloomFilter(Functions const &);
+  ~BasicBloomFilter() override;
+
+  BasicBloomFilter(BasicBloomFilter const &) = delete;
+  BasicBloomFilter(BasicBloomFilter &&)      = delete;
+  BasicBloomFilter &operator=(BasicBloomFilter const &) = delete;
+  BasicBloomFilter &operator=(BasicBloomFilter &&) = delete;
+
+  bool Match(Bytes const &) override;
+  void Add(Bytes const &) override;
+  bool ReportFalsePositives(std::size_t) override;
+
+public:
+  BitVector                         bits_;
+  internal::HashSourceFactory const hash_source_factory_;
+  std::size_t                       entry_count_{};
+  std::size_t                       positive_count_{};
+  std::size_t                       false_positive_count_{};
+};
+
+/*
+ * A fake Bloom filter which holds no data and treats any query as positive.
+ */
+class DummyBloomFilter : public BloomFilter
```

Comment:
As far as I can tell this is only used in tests therefore we should move it to the test folder.

Reply:
No: it's used by the main chain by default as long as the feature flag is not set, in which case the BasicBloomFilter is used instead